s3mapprovider for Collector: load configuration from config files in Amazon S3

junhaoyu-aws opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
An implementation of ConfigMapProvider for Amazon S3 (s3mapprovider) allows OTEL Collector the ability to load configuration for itself by fetching and reading config files stored in Amazon S3.

**Link to tracking Issue:** <Issue number if applicable>
#12939

**Testing:** <Describe what testing was performed and which tests were added.>
First, I mocked how a S3 client works while interacting with a S3 server.
Second, I designed and passed all the unit tests for edge cases such as empty fields, invalid config files, non-exist config files and so on.
Third, I did integration tests such as:
- Start an OTEL Collector with s3mapprovider.
- Start an OTEL sample application to send telemetry data to check if the OTEL Collector is running well.

**Documentation:** <Describe the documentation added.>
Added a README.md

**Sponsor:**
@Aneurysm9

rice-junhaoyu committed:
source codes for s3mapprovider

rice-junhaoyu committed:
changes for github checks

junhaoyu-aws committed:
Merge pull request #3 from junhaoyu-aws/s3upstream

s3mapprovider for Collector: load configuration from config files in Amazon S3

mx-psi commented on .github/ALLOWLIST:
```diff
@@ -14,6 +14,7 @@
 #
 
 cmd/configschema
+confmap/provider/s3mapprovider
```
Any new component or folder must have an owner. You should add yourself and @Aneurysm9  if he is sponsoring

```suggestion
confmap/provider/s3mapprovider
```

junhaoyu-aws replied:
Fixed.

mx-psi commented on confmap/provider/s3mapprovider/provider.go:
```diff
@@ -0,0 +1,135 @@
+	return &provider{client: s3.NewFromConfig(cfg)}
+}
+
+func (fmp *provider) Retrieve(ctx context.Context, uri string, _ confmap.WatcherFunc) (confmap.Retrieved, error) {
```
This will need an update because of open-telemetry/opentelemetry-collector/pull/5839

junhaoyu-aws replied:
Fixed, thanks for letting me know.

junhaoyu-aws committed:
Merge branch 'open-telemetry:main' into main

rice-junhaoyu committed:
CODEOWNERS add

junhaoyu-aws committed:
Merge pull request #4 from junhaoyu-aws/s3upstream

CODEOWNERS add

rice-junhaoyu committed:
Fixed return types - keep consistency according to the Break Change : Change confmap.Provider to return pointer to Retrieved

junhaoyu-aws committed:
Fixed return types for s3mapprovider - keep consistency according to the Break Change #5839 in upstream

Fixed return types for s3mapprovider - keep consistency according to the Break Change #5839 in upstream

mx-psi commented on .github/ALLOWLIST:
```diff
@@ -14,6 +14,7 @@
 #
 
 cmd/configschema
+confmap/provider/s3mapprovider
```
This is no longer needed
```suggestion
```

junhaoyu-aws replied:
Fixed!

mx-psi commented:
@Aneurysm9 to review

rice-junhaoyu committed:
Updated ALLOWLIST file, No longer need to specify

junhaoyu-aws committed:
Updated ALLOWLIST file, No longer need to specify

Updated ALLOWLIST file, No longer need to specify

Aneurysm9 commented on confmap/provider/s3mapprovider/README.md:
```diff
@@ -0,0 +1,14 @@
+What is this new component s3mapprovider?
+- An implementation of ConfigMapProvider for Amazon S3 (s3mapprovider) allows OTEL Collector the ability to load configuration for itself by fetching and reading config files stored in Amazon S3.
+
+How this new component s3mapprovider works?
+- It will be called by ConfigMapResolver to load configurations for OTEL Collector.
+- By giving a config URI starting with prefix 's3://', this s3mapprovider will be used to download config files from given S3 URIs, and then used the downloaded config files to deploy the OTEL Collector.
+- In our code, we check the validity scheme and string pattern of S3 URIs. And also check if there are any problems on config downloading and config deserialization.
```
```suggestion
## Summary

This package provides a `ConfigMapProvider` implementation for Amazon S3 (`s3mapprovider`) that allows the  Collector the ability to load configuration by fetching and reading config objects stored in Amazon S3.

## How it works
- It will be called by `ConfigMapResolver` to load configuration for the Collector.
- By giving a config URI starting with prefix `s3://`, this `s3mapprovider` will be used to download config objects from the given S3 URIs, and then use the downloaded configuration during Collector initialization.
```

This component won't always be new, so we should avoid referring to it as such in this document.  Made some other tweaks to clarify the content.

junhaoyu-aws replied:
Fixed! Thanks!

rice-junhaoyu committed:
README updated

junhaoyu-aws committed:
README updated : Avoid referring to the component as such in this document

README updated : Avoid referring to the component as such in this document

mx-psi commented on confmap/provider/s3mapprovider/provider.go:
```diff
@@ -0,0 +1,135 @@
+// `s3://DOC-EXAMPLE-BUCKET.s3.us-west-2.amazonaws.com/photos/puppy.jpg` - (unix, windows)
+func New() confmap.Provider {
+	// initialize the client
+	cfg, err := config.LoadDefaultConfig(context.Background())
+	if err != nil {
+		log.Fatal(err)
+	}
```
We should not use `log.Fatal` here, we never use the `log` package and this error may therefore be missed since it will not appear alongside the rest of logs.

I would initialize the S3 client on the first call to `Retrieve` now, and propose allowing the `New` constructor to return an error on opentelemetry-collector.

rice-junhaoyu replied:
Fixed the comments of `log.Fatal`, and `initialize the S3 client on the first call to Retrieve now`.

But for `allowing the New constructor to return an error on opentelemetry-collector `, I think it would better start a new issue and a new PR?

rice-junhaoyu committed:
Not using log.Fatal and log package any more

junhaoyu-aws committed:
Not using log.Fatal and log package any more

Not using log.Fatal and log package any more

junhaoyu-aws committed:
Merge branch 'open-telemetry:main' into main

rice-junhaoyu committed:
Collector version from v0.57.2 to v0.58; Lazy Initialize the S3 client

rice-junhaoyu committed:
Fixed import groups to be in the proper order

jeromeinsf commented:
/remove stale

jpkrohling commented:
@mx-psi, is this still missing something, or can it be merged?

mx-psi commented on confmap/provider/s3mapprovider/provider.go:
```diff
@@ -0,0 +1,138 @@
+	if fmp.client == nil {
+		cfg, err := config.LoadDefaultConfig(context.Background())
+		if err != nil {
+			fmt.Println("Failed to load configurations to initialize an AWS SDK client, error: %w", err)
```
My only concern is that we are using `fmt.Println` here. We can't use the logger, since configuration loading happens before we have a logger, so I don't know if we can do any better. Any thoughts @jpkrohling ?

rapphil replied:
It feels what we really need is to return an error in case the configuration cannot be loaded: there is no point in trying to move on with the creation of a s3 client if the configuration has a problem.

mx-psi replied:
Oh, yeah, you are right, I misread the following line. This should just return  an error.

junhaoyu-aws replied:
Yes, thanks for the suggestions, you are right. Fixed yet!

mx-psi commented on confmap/provider/s3mapprovider/provider.go:
```diff
@@ -0,0 +1,138 @@
+	// initialize the s3 client in the first call of Retrieve
+	if fmp.client == nil {
+		cfg, err := config.LoadDefaultConfig(context.Background())
+		if err != nil {
+			fmt.Println("Failed to load configurations to initialize an AWS SDK client, error: %w", err)
```
```suggestion
		if err != nil {
			return nil, fmt.Errorf("Failed to load configurations to initialize an AWS SDK client, error: %w", err)
```

junhaoyu-aws replied:
Fixed

rice-junhaoyu committed:
return err when default config cannot be loaded

rice-junhaoyu committed:
collector version fixed

rice-junhaoyu committed:
make gotidy

mx-psi commented:
@rice-junhaoyu  @Aneurysm9 there are two things to fix for merging this:

1. Add a CODEOWNER entry for this provider (I assume this would be someone from AWS?)
2. Sign the CLA for the commits by @rice-junhaoyu

rice-junhaoyu commented:
> @rice-junhaoyu @Aneurysm9 there are two things to fix for merging this:
> 
> 1. Add a CODEOWNER entry for this provider (I assume this would be someone from AWS?)
> 2. Sign the CLA for the commits by @rice-junhaoyu

@mx-psi The @rice-junhaoyu has signed CLA yet. I think the current issue is related to my deactivated AWS email address. The commits are pushed when I was interning at AWS.

I referred to this [link](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user) and figured out I need to add the AWS email address in my GitHub account's emails. But the problem is I am not able to use my AWS email address any more. Also, the old commits probably seem not being linked any more according to the link.

Do you have any idea on this, or I can restart a PR for this s3provider?

mx-psi commented:
> > @rice-junhaoyu @Aneurysm9 there are two things to fix for merging this:
> > 
> > 1. Add a CODEOWNER entry for this provider (I assume this would be someone from AWS?)
> > 2. Sign the CLA for the commits by @rice-junhaoyu
> 
> @mx-psi The @rice-junhaoyu has signed CLA yet. I think the current issue is related to my deactivated AWS email address. The commits are pushed when I was interning at AWS.

Ah, you are right, got it!

> I referred to this [link](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user) and figured out I need to add the AWS email address in my GitHub account's emails. But the problem is I am not able to use my AWS email address any more. Also, the old commits probably seem not being linked any more according to the link.
> 
> Do you have any idea on this, or I can restart a PR for this s3provider?

The nuclear option is to  
1. run `git rebase -i $(git merge-base main HEAD)`, 
2. squash all commits from your previous account into a single one (resulting in a new commit done by your new account). This should be a matter of following the instructions for squash in the editor screen that will come up.
3. Finish the rebase (you may need to do `git rebase --continue`)
4. Force push into this branch (`git push --force-with-lease`)

If that sounds too hard, opening a new PR is also an option of course :)

rice-junhaoyu commented:
@mx-psi  I tried to squash all commits onto my first commit c974ece, which is a single commit. But there are about 100 commits included when I called `git rebase -i c974ece`, maybe since I have merged latest commits from upstream to my fork before. On my understanding, this will lead to a case that those unrelated commits are also included in the file changes of this PR? That will be really ugly. :( 

If I am right, would you recommend me to open a new PR instead?

mx-psi commented:
> If I am right, would you recommend me to open a new PR instead?

@rice-junhaoyu Yes, new PR also works, please ping me and mention this PR so that people understand this code has already been reviewed. Thanks for taking the time to make this change and sorry about the hassle

rice-junhaoyu commented:
> > If I am right, would you recommend me to open a new PR instead?
> 
> @rice-junhaoyu Yes, new PR also works, please ping me and mention this PR so that people understand this code has already been reviewed. Thanks for taking the time to make this change and sorry about the hassle

@mx-psi Thanks for your understanding. So sorry about this, I didn't expect my deactivated AWS email address would lead to this situation. 

I've already opened a new PR for this and ping you yet.

codeboten commented:
Replaced by https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/14317
